<launch>
  <!-- Args to enable each sensor -->
  <arg name="use_lidar_center" default="true" />
  <arg name="use_lidar_center_image" default="false" />
  <arg name="lidar_min_dist" default="0.4" />
  <arg name="use_lidar_front" default="true" />
  <arg name="lidar_front_min_dist" default="0.4" />
  <arg name="use_camera_fl" default="false" />
  <arg name="use_camera_fr" default="false" />
  <arg name="use_radar_fc" default="false" />
  <arg name="use_gps" default="false" />
  <arg name="use_dbw" default="false" />
  <arg name="use_ssc" default="false" />
  <arg name="use_tf" default="false" />
  <arg name="use_diagnostics" default="true" />
  <arg name="gps_driver_name" default="novatel" />

  <!-- LIDAR Center - Ouster -->
  <group if="$(arg use_lidar_center)" ns="lidar_center">
    <include file="$(find vehicle_platform)/launch/ouster_lidar_driver.launch">
      <arg name="sensor_hostname" value="192.168.3.200" />
      <arg name="udp_dest" value="192.168.3.100" />
      <arg name="lidar_mode" value="1024x10" />
      <arg name="image" value="$(arg use_lidar_center_image)" />
      <arg name="tf_prefix" value="lidar_center" />
    </include>
  </group>

  <!-- LIDAR Front - Velodyne -->
  <group if="$(arg use_lidar_front)" ns="lidar_front">
    <remap from="velodyne_points" to="points_raw" />

    <include if="$(arg use_lidar_front)" file="$(find velodyne_pointcloud)/launch/VLP16_points.launch">
      <arg name="frame_id" value="lidar_front" />
      <arg name="device_ip" value="192.168.1.201" />
      <arg name="min_range" value="$(arg lidar_front_min_dist)" />
      <arg name="port" value="2368" />
    </include>
  </group>

  <!-- CAMERAS -->
  <!-- Front Left Camera -->
  <include if="$(arg use_camera_fl)" file="$(find vehicle_platform)/launch/avt_vimba_camera.launch">
    <arg name="name" value="camera_fl" />
    <arg name="frame_id" value="camera_fl" />
    <arg name="ip" value="192.168.20.1" />
    <arg name="acquisition_rate" value="10" />
  </include>

  <!-- Front Right Camera -->
  <include if="$(arg use_camera_fr)" file="$(find vehicle_platform)/launch/avt_vimba_camera.launch">
    <arg name="name" value="camera_fr" />
    <arg name="frame_id" value="camera_fr" />
    <arg name="ip" value="192.168.10.1" />
    <arg name="acquisition_rate" value="10" />
  </include>

  <!-- RADARS -->
  <!-- Front Center ESR  -->
  <group if="$(arg use_radar_fc)" ns="radar_fc">
    <include file="$(find delphi_esr)/launch/delphi_esr.launch">
      <arg name="esr_frame_id" value="radar_fc" />
      <arg name="viz_mature_tracks_only" value="true" />
      <arg name="use_kvaser" value="true" />
      <arg name="kvaser_hardware_id" value="10761" />
      <arg name="kvaser_circuit_id" value="1" />
    </include>
    <node pkg="vehicle_platform" name="radar_topic_merger" type="radar_topic_merger" />
  </group>

  <!-- GPS -->
  <group>
    <node if="$(arg use_diagnostics)" pkg="vehicle_platform" type="diagnostics_parser.py" name="diagnostics_parser" />
    <remap from="/diagnostics" to="/unprocessed_diagnostics" />
    <include if="$(arg use_gps)" file="$(find vehicle_platform)/launch/novatel_gps_driver.launch">
      <arg name="ip" value="192.168.100.150" />
      <arg name="port" value="2000" />
      <arg name="imu_rate" value="100" />
      <arg name="driver_name" value="$(arg gps_driver_name)" />
    </include>
  </group>

  <!-- DRIVE BY WIRE -->
  <include if="$(arg use_dbw)" file="$(find pacmod3)/launch/pacmod3.launch" ns="pacmod">
    <arg name="pacmod_vehicle_type" value="LEXUS_RX_450H" />
    <arg name="use_kvaser" value="true" />
    <arg name="kvaser_hardware_id" value="10761" />
    <arg name="kvaser_circuit_id" value="0" />
  </include>

  <!-- SSC -->
  <include if="$(arg use_ssc)" file="$(find ssc_pm_lexus)/launch/speed_steering_control.launch" >
    <arg name="speed_model_json" value="$(find vehicle_platform)/config/ssc/speed_model.json" />
    <arg name="steering_model_json" value="$(find vehicle_platform)/config/ssc/steering_model.json" />
    <arg name="veh_controller_json" value="$(find vehicle_platform)/config/ssc/veh_controller.json" />
    <arg name="veh_interface_json" value="$(find vehicle_platform)/config/ssc/veh_interface.json" />
  </include>

  <!-- TRANSFORMS -->
  <node if="$(arg use_tf)" pkg="robot_state_publisher" type="robot_state_publisher" name="vehicle_platform_frames">
    <remap from="robot_description" to="vehicle_platform_frames/robot_description" />
    <param name="robot_description" textfile="$(find vehicle_platform)/config/platform.urdf" />
  </node>

  <!-- DIAGNOSTICS -->
  <include if="$(arg use_diagnostics)" file="$(find vehicle_platform)/launch/diagnostics.launch" >
    <arg name="use_lidar_center" value="$(arg use_lidar_center)" />
    <arg name="use_lidar_front" value="$(arg use_lidar_front)" />
    <arg name="use_camera_fl" value="$(arg use_camera_fl)" />
    <arg name="use_camera_fr" value="$(arg use_camera_fr)" />
    <arg name="use_gps" value="$(arg use_gps)" />
  </include>
</launch>
